src/gluonnlp/embedding/token_embedding.py

```diff
@@ -1020,9 +1020,100 @@ class Word2Vec(TokenEmbedding):
+            for line_num in range(vocab_size):
+                token = []
+                while True:
+                    ch = f.read(1)
+                    if ch == b' ':
+                        break
+                    if ch == b'':
+                        raise EOFError('unexpected end of input; is count incorrect or file '
+                                       'otherwise damaged?')
+                    if ch != b'\n':  # ignore newlines in front of words (some binary files have)
+                        token.append(ch)
+                try:
+                    token = b''.join(token).decode(encoding=encoding)
+                except ValueError:
+                    warnings.warn('line {} in {}: failed to decode. Skipping.'
+                                  .format(line_num, pretrained_file_path))
+                    continue
+                elems = np.fromstring(f.read(binary_len), dtype=np.float32)
+
+                assert len(elems) > 1, 'line {} in {}: unexpected data format.'.format(
+                    line_num, pretrained_file_path)
+
+                if token == self.unknown_token and loaded_unknown_vec is None:
+                    loaded_unknown_vec = elems
+                    tokens.add(self.unknown_token)
+                elif token in tokens:
+                    warnings.warn('line {} in {}: duplicate embedding found for '
+                                  'token "{}". Skipped.'.format(line_num, pretrained_file_path,
+                                                                token))
+                else:
+                    assert len(elems) == vec_len, \
+                        'line {} in {}: found vector of inconsistent dimension for token ' \
+                        '"{}". expected dim: {}, found: {}'.format(line_num,
+                                                                   pretrained_file_path,
+                                                                   token, vec_len, len(elems))
+                    all_elems.extend(elems)
+                    self._idx_to_token.append(token)
+                    self._token_to_idx[token] = len(self._idx_to_token) - 1
+                    tokens.add(token)
+        self._idx_to_vec = nd.array(all_elems).reshape((-1, vec_len))
+
+        if self.unknown_token:
+            if loaded_unknown_vec is None:
+                self._idx_to_vec[C.UNK_IDX] = self._init_unknown_vec(shape=vec_len)
+            else:
+                self._idx_to_vec[C.UNK_IDX] = nd.array(loaded_unknown_vec)
+
+    @classmethod
+    def from_binary(cls, pretrained_file_path, encoding='utf8'):
+        """Load embedding vectors from a binary pre-trained token embedding file.
+
+        Parameters
+        ----------
+        pretrained_file_path: str
+            The path to a binary pre-trained token embedding file.
+        encoding: str
+            The encoding type of the file.
+        """
+        embedding = cls(None)
```

Comment:
I want to discuss about this line. Should I use `embedding = TokenEmbedding(**kwargs)` or current implementation.

Reply:
I think this should probably be named as from_w2v_binary, as the binary format is different from fasttext binary. Once it has this naming, then returning Word2Vec embedding object makes sense.

Reply:
renamed in a4e5f6a

Reply:
The current implementation requires that creation of an empty `Word2Vec` object is possible, for which there is actually no use besides the private `_load_w2v_binary`. Switching to `embedding = TokenEmbedding(**kwargs)` results in initializing / returning a parent class object instead of `Word2Vec`. Both seems ok, but not ideal. I'm ok to keep it. Alternatively, taking the `source` argument of `__init__` to mean either the name (current functionality) or path to binary could be possible?

My suggestion for a separate `from_binary` initializer was based on the `model.train.embedding` https://github.com/dmlc/gluon-nlp/blob/d49a789/src/gluonnlp/model/train/embedding.py#L232. In that case, there was no `source` argument whose meaning could be overloaded, and introducing one wouldn't make sense either as it would conflict with other initializer arguments. Here the case is different. Sorry for not pointing this out before.

What do you think?

Reply:
@leezu thanks for your review. I agree with you either using overload or parent object is not ideal. However, I might not fully understand what you really want to implement it. I will try to explain what I have thoughts. 

Do you mean the constructor can be implemented like this? However, the error would be unclear for users, right?
```py
class Word2Vec(TokenEmbedding):
    source_file_hash = C.WORD2VEC_NPZ_SHA1
    def __init__(self, source='GoogleNews-vectors-negative300',
                 embedding_root=os.path.join(get_home_dir(), 'embedding'), encoding='utf8',
                 **kwargs):
        super(Word2Vec, self).__init__(**kwargs)
        try:
            self._check_source(self.source_file_hash, source)
            pretrained_file_path = self._get_file_path(self.source_file_hash,
                                                       embedding_root, source)
            self._load_embedding(pretrained_file_path, elem_delim=' ')
        except KeyError:
            pretrained_file_path = os.path.expanduser(source)
            if is_binary(pretrained_file_path): 
                self._load_w2v_binary(pretrained_file_path, encoding=encoding)
            raise TypeError('{} is not a binary file'.format(pretrained_file_path))
```

Reply:
Yes, I meant such `__init__` as in your example. Do you think it is more appropriate than introducing a second constructor?

To clarify the error message: If source neither refers to a valid name for `self._load_embedding` nor to a valid path (ie. `except KeyError`), raise a generic error that `source` must be either a valid name or a path to binary file.

Only `raise TypeError('{} is not a binary file'.format(pretrained_file_path))` if `source` is not a valid name, but a valid path to an existing file, which however is not of the correct binary format.

What do you think?

Reply:
@leezu Thanks. 
The code has been updated, but this implementation doesn't have the `is_binary`. I decide to remove it because there is no a really good way to check a binary file in python. And I think if the file is not in a good format, errors will be raised in `_load_w2v_binary`. 
For text file, it is loaded from `from_file`, so it doesn't go through this constructor.  

Does it make sense to you?

Reply:
Let's stick to a convention for binary files. I think an extension `.bin` is good enough for representing it for now.